doc/neps/nep-0050-scalar-promotion.rst

```diff
@@ -0,0 +1,743 @@
+     - *unchanged*
+   * - ``array([0.1], float32) == float64(0.1)``
+     - ``array([ True])``
+     - ``array([False])``  [T7]_
+   * - ``array([1.], float32) + 3``
+     - ``array([4.], float32)``
+     - *unchanged*
+   * - ``array([1.], float32) + int64(3)``
+     - ``array([4.], float32)``
+     - ``array([4.], float64)``  [T8]_
+
+.. [T1] New behaviour honours the dtype of the ``uint8`` scalar.
+.. [T2] Current NumPy ignores the precision of 0-D arrays or NumPy scalars
+        when combined with arrays.
+.. [T3] Current NumPy ignores the precision of 0-D arrays or NumPy scalars
+        when combined with arrays.
+.. [T4] Old behaviour uses ``uint16`` because ``300`` does not fit ``uint8``,
+        new behaviour raises an error for the same reason.
+.. [T5] ``300`` cannot be converted to ``uint8``.
+.. [T6] ``np.float32(3e100)`` overflows to infinity.
+.. [T7] ``0.1`` loses precision when cast to ``float32``, but old behaviour
+        casts the ``float64(0.1)`` and then matches.
+.. [T8] NumPy promotes ``float32`` and ``int64`` to ``float64``.  The old
+        behaviour ignored the ``int64`` here.
+
+
+Motivation and Scope
+====================
+
+The motivation for changing the behaviour with respect to inspecting the value
+of Python scalars and NumPy scalars/0-D arrays is two-fold:
+
+1. The special handling of NumPy scalars/0-D arrays as well as the value
+   inspection can be very surprising to users,
+2. The value-inspection logic is much harder to explain and implement.
+   It is further harder to make it available to user defined DTypes through
+   :ref:`NEP 42 <NEP42>`.
+   Currently, this leads to a dual implementation of a new and an old (value
+   sensitive) system.  Fixing this will greatly simplify the internal logic
+   and make results more consistent.
+
+We believe that the proposal of "weak" Python scalars will help users by
+providing a clear mental model for which datatype an operation will
+result in.
+This model fits well with the preservation of array precisions that NumPy
+currently often follows, and also uses for in-place operations::
+
+    arr += value
+
+Preserves precision as long as "kind" boundaries are not crossed (otherwise
+an error is raised).
+
+While some users will potentially miss the value inspecting behavior even for
+those cases where it seems useful, it quickly leads to surprises.  This may be
+expected::
+
+    np.array([100], dtype=np.uint8) + 1000 == np.array([1100], dtype=np.uint16)
+
+But the following will then be a surprise::
+
+    np.array([100], dtype=np.uint8) + 200 == np.array([44], dtype=np.uint8)
+
+Considering that the proposal aligns with the behavior of in-place operands
+and avoids the surprising switch in behavior that only sometimes avoids
+overflow in the result,
+we believe that the proposal follows the "principle of least surprise".
+
+
+Usage and Impact
+================
+
+This NEP is expected to be implemented with **no** transition period that warns
```

Comment:
Yikes; this has the potential to break some codes out there, without any obvious way to fix them.  We should probably, at least, make a major version release.